lib/streamlit/elements/exception.py

```diff
@@ -218,36 +201,80 @@ def _get_stack_trace_str_list(
     if extracted_traceback is None:
-        stack_trace_str_list = [
+        trace_str_list = [
             "Cannot extract the stack trace for this exception. "
             "Try calling exception() within the `catch` block."
         ]
     else:
-        if strip_streamlit_stack_entries:
-            extracted_frames = _get_nonstreamlit_traceback(extracted_traceback)
-            stack_trace_str_list = traceback.format_list(extracted_frames)
+        internal_frames, external_frames = _split_internal_streamlit_frames(
+            extracted_traceback
+        )
+
+        if external_frames:
+            trace_str_list = traceback.format_list(external_frames)
         else:
-            stack_trace_str_list = traceback.format_list(extracted_traceback)
+            trace_str_list = traceback.format_list(internal_frames)
 
-    stack_trace_str_list = [item.strip() for item in stack_trace_str_list]
+        trace_str_list = [item.strip() for item in trace_str_list]
 
-    return stack_trace_str_list
+    return trace_str_list
 
 
-def _is_in_streamlit_package(file: str) -> bool:
-    """True if the given file is part of the streamlit package."""
+def _is_in_package(file: str, package_path: str) -> bool:
+    """True if the given file is part of package_path."""
     try:
-        common_prefix = os.path.commonprefix([os.path.realpath(file), _STREAMLIT_DIR])
+        common_prefix = os.path.commonprefix([os.path.realpath(file), package_path])
     except ValueError:
         # Raised if paths are on different drives.
         return False
 
-    return common_prefix == _STREAMLIT_DIR
+    return common_prefix == package_path
 
 
-def _get_nonstreamlit_traceback(
+def _split_internal_streamlit_frames(
     extracted_tb: traceback.StackSummary,
-) -> list[traceback.FrameSummary]:
-    return [
-        entry for entry in extracted_tb if not _is_in_streamlit_package(entry.filename)
-    ]
+) -> tuple[list[traceback.FrameSummary], list[traceback.FrameSummary]]:
+    """Split the traceback into a Streamlit-internal part and an external part.
+
+    The internal part is everything up to (but excluding) the first frame belonging to
+    the user's code. The external part is everything else.
+
+    So if the stack looks like this:
+
+        1. Streamlit frame
+        2. Pandas frame
+        3. Altair frame
+        4. Streamlit frame
+        5. User frame
+        6. User frame
+        7. Streamlit frame
+        8. Matplotlib frame
+
+    ...then this should return 1-4 as the internal traceback and 5-8 as the external.
+
+    (Note that something like the example above is extremely unlikely to happen since
+    it's not like Altair is calling Streamlit code, but you get the idea.)
+    """
+
+    ctx = get_script_run_ctx()
+
+    if not ctx:
+        return [], list(extracted_tb)
+
+    package_path = os.path.join(os.path.realpath(str(ctx.main_script_parent)), "")
+
+    internal: list[traceback.FrameSummary] = []
```

Comment:
One idea I had while reading this is that it could be nice to extract the logic starting here into a separate function. Then we could have a unit test with dummy extracted traceback data. 

Looking through your tests, I am on the fence about whether it is necessary to test on this more granular level.

Reply:
I see what you're saying. We could split 266:280 into a `_partition_list()` function that we use like this:

```py
internal, external = partition_list(extracted_tb, find_first_occurrence_of=lambda fr: _is_in_package(fr.filename, package_path))
```

Though, yeah, I'm also on the fence!

Reply:
Ok I did what you proposed and it's actually super nice. Thanks :)